src/utils/stats_alloc/src/task_local.rs

```diff
@@ -0,0 +1,203 @@
+            // reach zero eventually and then `drop` itself.
+            let _guard = Box::new(114514);
+            let monitor = async move {
+                let mut interval = tokio::time::interval(interval);
+                loop {
+                    interval.tick().await;
+                    BYTES_ALLOCATED.with(|bytes| report(bytes.val()));
+                }
+            };
+            let output = tokio::select! {
+                biased;
+                _ = monitor => unreachable!(),
+                output = future => output,
+            };
+            output
+        })
+        .await
+}
+
+#[inline(always)]
+fn wrap_layout(layout: Layout) -> (Layout, usize) {
+    debug_assert_ne!(layout.size(), 0, "the size of layout must be non-zero");
+
+    let (wrapped_layout, offset) = Layout::new::<TaskLocalBytesAllocated>()
+        .extend(layout)
+        .expect("wrapping layout overflow");
+    let wrapped_layout = wrapped_layout.pad_to_align();
+
+    (wrapped_layout, offset)
+}
+
+pub struct TaskLocalAlloc<A>(pub A);
+
+unsafe impl<A> GlobalAlloc for TaskLocalAlloc<A>
+where
+    A: GlobalAlloc,
+{
+    unsafe fn alloc(&self, layout: Layout) -> *mut u8 {
+        let (wrapped_layout, offset) = wrap_layout(layout);
+
+        BYTES_ALLOCATED
+            .try_with(|&bytes| {
+                bytes.add_unchecked(layout.size());
+                let ptr = self.0.alloc(wrapped_layout);
+                *ptr.cast() = bytes;
+                ptr.wrapping_add(offset)
+            })
+            .unwrap_or_else(|_| {
+                let ptr = self.0.alloc(wrapped_layout);
+                *ptr.cast() = TaskLocalBytesAllocated::invalid();
+                ptr.wrapping_add(offset)
+            })
+    }
+
+    unsafe fn dealloc(&self, ptr: *mut u8, layout: Layout) {
+        let (wrapped_layout, offset) = wrap_layout(layout);
+        let ptr = ptr.wrapping_sub(offset);
+
+        let bytes: TaskLocalBytesAllocated = *ptr.cast();
+        bytes.sub(layout.size());
+
+        self.0.dealloc(ptr, wrapped_layout);
+    }
+
+    unsafe fn alloc_zeroed(&self, layout: Layout) -> *mut u8 {
+        let (wrapped_layout, offset) = wrap_layout(layout);
+
+        BYTES_ALLOCATED
+            .try_with(|&bytes| {
+                bytes.add_unchecked(layout.size());
+                let ptr = self.0.alloc_zeroed(wrapped_layout);
+                *ptr.cast() = bytes;
+                ptr.wrapping_add(offset)
+            })
+            .unwrap_or_else(|_| {
+                let ptr = self.0.alloc_zeroed(wrapped_layout);
+                *ptr.cast() = TaskLocalBytesAllocated::invalid();
+                ptr.wrapping_add(offset)
+            })
+    }
+
+    unsafe fn realloc(&self, ptr: *mut u8, layout: Layout, new_size: usize) -> *mut u8 {
+        let (wrapped_layout, offset) = wrap_layout(layout);
+        let ptr = ptr.wrapping_sub(offset);
+
+        let bytes: TaskLocalBytesAllocated = *ptr.cast();
+        bytes.add(new_size);
```

Comment:
Shouldn't we apply this to the task local var instead?

Reply:
As above, the deallocation should always be associated with the task that originally allocates it, so we don't use task local var here.

Reply:
so, how do we tally up the total deallocated/reallocated?

It doesn't seem to be done here.

Reply:
oh, I see, we use a static ptr.

Anw, I'd still like to see a unittest, including multithreaded test, that shows this works as expected (including across threads).